ENH: Added missing methods to `np.flatiter`

BvB93 opened the pull request:
Closes https://github.com/numpy/numpy/issues/17114 and https://github.com/numpy/numpy/issues/17113.

This pull requests adds annotations to the following three `np.flatiter` methods:

* `__getitem__()`
* `__len__()`
* `__array__()`

Commit:
MAINT: Added missing methods to `np.flatiter`

Added `__len__`, `__getitem__` and `__array__`

Commit:
MAINT: Replace `_SupportsArray` with `ndarray`

Commit:
TST: Added typing tests for `np.flatiter`

Commit:
TST: Added an `__array__` test

Commit:
TST,MAINT: Fixed a few tests

BvB93 commented:
Pinging @bashtage.

Commit:
TST: Added a missing function call

eric-wieser commented on numpy/tests/typing/fail/flatiter.py:
```diff
@@ -0,0 +1,21 @@
+a.index = Any  # E: Property "index" defined in "flatiter" is read-only
+a.copy(order='C')  # E: Unexpected keyword argument
+a[np.bool_()]  # E: No overload variant of "__getitem__"
+a[Index()]  # E: No overload variant of "__getitem__"
```
Shouldn't this be legal? Might be worth referencing an issue if its not.

BvB93 replied:
It should be, but the behavior of `flatiter.__getitem__` is very much inconsistent 
with respect to `ndarray.__getitem__` (case in point: https://github.com/numpy/numpy/issues/17175).

Adding a note is probably a good idea though (see https://github.com/numpy/numpy/pull/17180/commits/edcf0ea581546eb8c29ba34b7576f2a8d81b4eec).

Commit:
DOC: Added a comment about weird `flatiter.__getitem__` behavior

Contrary to `ndarray.__getitem__` its counterpart in `flatiter` does not accept objects with the `__array__` or `__index__` protocols;
boolean indexing is just plain broken (gh-17175)

bashtage commented on numpy/__init__.pyi:
```diff
@@ -143,6 +143,14 @@ class _flagsobj:
     def __setitem__(self, key: str, value: bool) -> None: ...
 
+_ArrayLikeInt = Union[
+    int,
```
Does [`int`, `integer`] include `bool`?

BvB93 replied:
It does as `bool` is a subclass of `int`.

There is, as far as I'm aware, no convenient way of annotating a function which accepts integers but rejects booleans.

eric-wieser reviewed (approved):
Looks safe enough

mattip commented:
Thanks @BvB93
